examples/sampling/graphbolt/link_prediction.py

```diff
@@ -0,0 +1,241 @@
+        data_block_converter,
+        args.neg_ratio,
+        format,
+        graph,
+    )
+    subgraph_sampler = gb.NeighborSampler(negative_sampler, graph, args.fanout)
+    feature_keys = [("node", None, "feat")]
+    feature_fetcher = gb.FeatureFetcher(
+        subgraph_sampler, features, feature_keys
+    )
+    dataloader = gb.MultiProcessDataLoader(
+        gb.CopyTo(feature_fetcher, torch.device("cpu")),
+        num_workers=args.num_workers,
+    )
+    return dataloader
+
+
+def to_dgl_blocks(sampled_subgraphs):
+    """Convert sampled subgraphs to DGL blocks."""
+    blocks = [
+        dgl.create_block(
+            sampled_subgraph.node_pairs,
+            num_src_nodes=sampled_subgraph.reverse_column_node_ids.shape[0],
+            num_dst_nodes=sampled_subgraph.reverse_row_node_ids.shape[0],
+        )
+        for sampled_subgraph in sampled_subgraphs
+    ]
+    return blocks
+
+
+@torch.no_grad()
+def evaluate(args, graph, features, current_set, model):
+    evaluator = Evaluator(name="ogbl-citation2")
+
+    # Since we need to evaluate the model, we need to set the number
+    # of layers to 1 and the fanout to -1.
+    args.fanout = [torch.LongTensor([-1])]
+    dataloader = get_dataloader(
+        args, graph, features, current_set, is_train=False
+    )
+    pos_pred = []
+    neg_pred = []
+
+    model.eval()
+    for step, data in tqdm.tqdm(enumerate(dataloader)):
+        # Unpack LinkPredictionBlock.
+        compacted_pairs = data.compacted_node_pair
+        node_feature = data.node_feature[(None, "feat")].float()
+        blocks = to_dgl_blocks(data.sampled_subgraphs)
+
+        # Get the embeddings of the input nodes.
+        y = model(blocks, node_feature)
+        pos_score = (
+            (model.predictor(y[compacted_pairs[0]] * y[compacted_pairs[1]]))
+            .squeeze()
+            .detach()
+        )
+        # We need to repeat the negative source nodes to match the number of
+        # negative destination nodes.
+        neg_src = compacted_pairs[0].repeat_interleave(args.neg_ratio, dim=0)
+        neg_score = (
+            model.predictor(y[neg_src] * y[data.compacted_negative_tail])
+            .squeeze()
+            .detach()
+        )
+        pos_pred.append(pos_score)
+        neg_pred.append(neg_score)
+    pos_pred = torch.cat(pos_pred, dim=0)
+    neg_pred = torch.cat(neg_pred, dim=0).view(-1, args.neg_ratio)
+
+    input_dict = {"y_pred_pos": pos_pred, "y_pred_neg": neg_pred}
+    mrr = evaluator.eval(input_dict)["mrr_list"]
+    return mrr, mrr.mean()
+
+
+def train(args, graph, features, train_set, valid_set, model):
+    total_loss = 0.0
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
+    dataloader = get_dataloader(args, graph, features, train_set)
+
+    for epoch in tqdm.trange(args.epochs):
+        model.train()
+        for step, data in enumerate(dataloader):
+            # Unpack LinkPredictionBlock.
+            labels = data.label.float()
```

Comment:
where is `data.label` is assigned? data_block_converter?

Reply:
In `NegativeSampler._collate` https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/negative_sampler.py#L117-L121.